File: android/src/test/kotlin/com/compress/all/flutter_compress/SizeMathTest.kt
package com.compress.all.flutter_compress

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * Boundary coverage for the Kotlin half of the three-way SizeMath port
 * (CLAUDE.md §12.2).
 *
 * These cases intentionally mirror `test/size_math_test.dart` one for one: the
 * algorithm is duplicated across Kotlin, Swift and Dart, so the only way to
 * catch drift is to assert the same numbers on each side. If you change a case
 * here, change it there too.
 */
class SizeMathTest {

    private fun config(
        quality: String = "medium",
        qualityPercent: Int? = null,
        targetSizeMB: Int? = null,
        videoBitrateKbps: Int? = null,
        maxWidth: Int? = null,
        maxHeight: Int? = null,
        removeAudio: Boolean = false,
        audioBitrateKbps: Int? = null,
        alignment: String = "auto16",
    ) = CompressionConfig(
        quality = quality,
        qualityPercent = qualityPercent,
        targetSizeMB = targetSizeMB,
        videoBitrateKbps = videoBitrateKbps,
        codec = "h265",
        maxWidth = maxWidth,
        maxHeight = maxHeight,
        frameRate = null,
        removeAudio = removeAudio,
        audioBitrateKbps = audioBitrateKbps,
        trimStartMs = null,
        trimEndMs = null,
        alignment = alignment,
        keepOriginalIfLarger = true,
        minSavingsPercent = 0,
        notification = null,
    )

    private fun bitrate(
        config: CompressionConfig,
        durationMs: Long = 60_000,
        sourceBitrateKbps: Int = 9_000,
        targetHeight: Int = 1080,
    ) = SizeMath.videoBitrateBps(config, durationMs, sourceBitrateKbps, targetHeight)

    // ---- priority chain ----------------------------------------------------

    @Test
    fun `targetSizeMB wins over every other control`() {
        val bps = bitrate(
            config(targetSizeMB = 10, videoBitrateKbps = 99_999, qualityPercent = 10),
        )
        val expected = (10L * 8 * 1024 * 1024 * 0.95 - 128_000 * 60.0) / 60.0
        assertEquals(expected.toInt(), bps)
    }

    @Test
    fun `videoBitrateKbps wins over quality controls`() {
        assertEquals(
            2_500_000,
            bitrate(config(videoBitrateKbps = 2_500, qualityPercent = 10)),
        )
    }

    @Test
    fun `qualityPercent wins over the preset tier`() {
        assertEquals(
            9_000 * 1000 * 25 / 100,
            bitrate(config(quality = "high", qualityPercent = 25)),
        )
    }

    @Test
    fun `preset tiers map to their documented percentages`() {
        val src = 10_000
        fun forTier(tier: String) = bitrate(config(quality = tier), sourceBitrateKbps = src)
        assertEquals(src * 1000 * 80 / 100, forTier("high"))
        assertEquals(src * 1000 * 50 / 100, forTier("medium"))
        assertEquals(src * 1000 * 30 / 100, forTier("low"))
        assertEquals(src * 1000 * 15 / 100, forTier("veryLow"))
    }

    // ---- edges -------------------------------------------------------------

    @Test
    fun `never re-inflates above the source bitrate`() {
        val bps = bitrate(config(qualityPercent = 100), sourceBitrateKbps = 4_000)
        assertTrue(bps <= 4_000 * 1000, "expected <= source, got $bps")
    }

    @Test
    fun `a source bitrate below the floor clamps up instead of failing`() {
        assertEquals(100_000, bitrate(config(), sourceBitrateKbps = 50))
    }

    @Test
    fun `an absurdly small target size still clears the floor`() {
        assertEquals(
            100_000,
            bitrate(config(targetSizeMB = 1), durationMs = 3_600_000),
        )
    }

    @Test
    fun `a zero-length source does not divide by zero`() {
        // coerceAtLeast(0.001) on the duration keeps this finite.
        assertTrue(bitrate(config(targetSizeMB = 10), durationMs = 0) > 0)
    }

    @Test
    fun `unknown source bitrate falls back to a resolution baseline`() {
        val bps = bitrate(config(), sourceBitrateKbps = 0, targetHeight = 720)
        val width = 720 * 16 / 9
        assertEquals((width * 720 * 30 * 0.12).toInt(), bps)
    }

    @Test
    fun `removeAudio hands the whole target budget to video`() {
        val withAudio = bitrate(config(targetSizeMB = 10))
        val without = bitrate(config(targetSizeMB = 10, removeAudio = true))
        assertTrue(without > withAudio, "$without should exceed $withAudio")
    }

    @Test
    fun `a custom audio bitrate changes the video budget`() {
        val low = bitrate(config(targetSizeMB = 10, audioBitrateKbps = 64))
        val high = bitrate(config(targetSizeMB = 10, audioBitrateKbps = 256))
        assertTrue(low > high)
    }

    // ---- dimensions --------------------------------------------------------

    @Test
    fun `no cap keeps the source exactly, without aligning`() {
        // Regression: auto16 used to round 1080 *up* to 1088, which broke
        // "downscale only" and forced a needless full-frame GL rescale.
        assertEquals(1080 to 2160, SizeMath.targetDimensions(1080, 2160, config()))
    }

    @Test
    fun `alignment never rounds up`() {
        val (w, _) = SizeMath.targetDimensions(2000, 1000, config(maxWidth = 1000))
        assertTrue(w <= 1000, "alignment upscaled to $w")
        assertEquals(0, w % 16)
    }

    @Test
    fun `only ever scales down - a cap above the source is a no-op`() {
        assertEquals(
            640 to 480,
            SizeMath.targetDimensions(640, 480, config(maxWidth = 4000, maxHeight = 4000)),
        )
    }

    @Test
    fun `caps each axis independently, preserving aspect`() {
        val (w, h) = SizeMath.targetDimensions(
            4000, 2000, config(maxWidth = 1920, alignment = "none"),
        )
        assertTrue(w <= 1920)
        assertTrue(kotlin.math.abs(w.toDouble() / h - 2) < 0.05, "aspect drifted: ${w}x$h")
    }

    @Test
    fun `alignment none still yields even dimensions`() {
        val (w, h) = SizeMath.targetDimensions(
            1001, 667, config(maxWidth = 1000, alignment = "none"),
        )
        assertEquals(0, w % 2)
        assertEquals(0, h % 2)
    }

    @Test
    fun `degenerate source dimensions pass through instead of throwing`() {
        assertEquals(0 to 0, SizeMath.targetDimensions(0, 0, config()))
        assertEquals(-1 to -1, SizeMath.targetDimensions(-1, -1, config(maxWidth = 100)))
    }

    @Test
    fun `never returns a zero side for a tiny cap`() {
        val (w, h) = SizeMath.targetDimensions(1920, 1080, config(maxWidth = 8))
        assertTrue(w > 0 && h > 0, "got ${w}x$h")
    }

    // ---- keepsOriginal (minSavingsPercent) ---------------------------------

    @Test
    fun `percent 0 keeps the original only when the output is not smaller`() {
        // The pre-threshold behaviour, preserved exactly.
        assertTrue(SizeMath.keepsOriginal(1000, 1000, true, 0))
        assertTrue(SizeMath.keepsOriginal(1001, 1000, true, 0))
        assertFalse(SizeMath.keepsOriginal(999, 1000, true, 0))
    }

    @Test
    fun `a threshold rejects marginal wins`() {
        // 5% of 1000 is 50, so anything at or above 950 is not worth swapping.
        assertTrue(SizeMath.keepsOriginal(970, 1000, true, 5))
        assertTrue(SizeMath.keepsOriginal(950, 1000, true, 5))
        assertFalse(SizeMath.keepsOriginal(949, 1000, true, 5))
    }

    @Test
    fun `the flag gates the whole behaviour`() {
        assertFalse(SizeMath.keepsOriginal(5000, 1000, false, 50))
    }

    @Test
    fun `an unknown source size never forces the original`() {
        // Web can report 0 when the blob size is unavailable; returning true
        // there would hand back an input we may not even be able to read.
        assertFalse(SizeMath.keepsOriginal(500, 0, true, 5))
    }
}

File: test/size_math_test.dart
// Boundary coverage for the shared size/bitrate math (CLAUDE.md §12.2).
//
// This is the algorithm all three platforms implement; a regression here is a
// regression everywhere, so the edges matter more than the happy path.

import 'package:flutter_compress_pro/flutter_compress_pro.dart';
import 'package:flutter_test/flutter_test.dart';

/// 60s at 24fps — a realistic clip length for the target-size arithmetic.
const _durationMs = 60000;

int _bitrate(
  VideoCompressConfig config, {
  int durationMs = _durationMs,
  int sourceBitrateKbps = 9000,
  int targetHeight = 1080,
  bool reserveAudio = true,
}) => SizeMath.videoBitrateBps(
  config: config,
  durationMs: durationMs,
  sourceBitrateKbps: sourceBitrateKbps,
  targetHeight: targetHeight,
  reserveAudio: reserveAudio,
);

void main() {
  group('videoBitrateBps priority chain', () {
    test('targetSizeMB wins over every other control', () {
      final bps = _bitrate(
        const VideoCompressConfig(
          targetSizeMB: 10,
          videoBitrateKbps: 99999,
          qualityPercent: 10,
          quality: CompressQuality.veryLow,
        ),
      );
      // 10 MiB × 8 × 0.95 − 128kbps×60s, over 60s.
      const expected = (10 * 8 * 1024 * 1024 * 0.95 - 128000 * 60) / 60;
      expect(bps, expected.toInt());
    });

    test('videoBitrateKbps wins over quality controls', () {
      expect(
        _bitrate(
          const VideoCompressConfig(videoBitrateKbps: 2500, qualityPercent: 10),
        ),
        2500 * 1000,
      );
    });

    test('qualityPercent wins over the preset tier', () {
      // 25% of a 9000 kbps source, not the 80% that `high` would give.
      expect(
        _bitrate(
          const VideoCompressConfig(
            qualityPercent: 25,
            quality: CompressQuality.high,
          ),
        ),
        9000 * 1000 * 25 ~/ 100,
      );
    });

    test('preset tiers map to their documented percentages', () {
      const srcKbps = 10000;
      int forTier(CompressQuality q) =>
          _bitrate(VideoCompressConfig(quality: q), sourceBitrateKbps: srcKbps);
      expect(forTier(CompressQuality.high), srcKbps * 1000 * 80 ~/ 100);
      expect(forTier(CompressQuality.medium), srcKbps * 1000 * 50 ~/ 100);
      expect(forTier(CompressQuality.low), srcKbps * 1000 * 30 ~/ 100);
      expect(forTier(CompressQuality.veryLow), srcKbps * 1000 * 15 ~/ 100);
    });
  });

  group('videoBitrateBps edges', () {
    test('never re-inflates above the source bitrate', () {
      // percent caps at 100, so the result can only ever be <= source.
      final bps = _bitrate(
        const VideoCompressConfig(qualityPercent: 100),
        sourceBitrateKbps: 4000,
      );
      expect(bps, lessThanOrEqualTo(4000 * 1000));
    });

    test('a source bitrate below the floor does not throw', () {
      // Regression: a clamp(floor, srcBps) here had lower > upper and threw
      // ArgumentError for any source under 100 kbps.
      expect(
        () => _bitrate(
          const VideoCompressConfig(quality: CompressQuality.medium),
          sourceBitrateKbps: 50,
        ),
        returnsNormally,
      );
      expect(
        _bitrate(
          const VideoCompressConfig(quality: CompressQuality.medium),
          sourceBitrateKbps: 50,
        ),
        SizeMath.minVideoBps,
      );
    });

    test('an absurdly small target size still clears the floor', () {
      expect(
        _bitrate(
          const VideoCompressConfig(targetSizeMB: 1),
          durationMs: 3600000,
        ),
        SizeMath.minVideoBps,
      );
    });

    test('a zero-length source does not divide by zero', () {
      expect(
        () => _bitrate(
          const VideoCompressConfig(targetSizeMB: 10),
          durationMs: 0,
        ),
        returnsNormally,
      );
    });

    test('unknown source bitrate falls back to a resolution baseline', () {
      final bps = _bitrate(
        const VideoCompressConfig(quality: CompressQuality.medium),
        sourceBitrateKbps: 0,
        targetHeight: 720,
      );
      // 1280x720 baseline at the "medium" anchor.
      const width = 720 * 16 ~/ 9;
      expect(bps, (width * 720 * 30 * 0.12).toInt());
    });

    test('removeAudio hands the whole target budget to video', () {
      final withAudio = _bitrate(const VideoCompressConfig(targetSizeMB: 10));
      final without = _bitrate(
        const VideoCompressConfig(targetSizeMB: 10, removeAudio: true),
      );
      expect(without, greaterThan(withAudio));
    });

    test('reserveAudio:false matches removeAudio (the web pipeline)', () {
      final web = _bitrate(
        const VideoCompressConfig(targetSizeMB: 10),
        reserveAudio: false,
      );
      final native = _bitrate(
        const VideoCompressConfig(targetSizeMB: 10, removeAudio: true),
      );
      expect(web, native);
    });

    test('a custom audio bitrate changes the video budget', () {
      final low = _bitrate(
        const VideoCompressConfig(targetSizeMB: 10, audioBitrateKbps: 64),
      );
      final high = _bitrate(
        const VideoCompressConfig(targetSizeMB: 10, audioBitrateKbps: 256),
      );
      expect(low, greaterThan(high));
    });
  });

  group('targetDimensions', () {
    test('no cap keeps the source exactly, without aligning', () {
      // Regression: auto16 used to round 1080 *up* to 1088, which both broke
      // "downscale only" and forced a needless full-frame rescale.
      expect(
        SizeMath.targetDimensions(1080, 2160, const VideoCompressConfig()),
        (1080, 2160),
      );
    });

    test('alignment never rounds up', () {
      // 1000 -> 992 (÷16), never 1008.
      final (w, _) = SizeMath.targetDimensions(
        2000,
        1000,
        const VideoCompressConfig(maxWidth: 1000),
      );
      expect(w, lessThanOrEqualTo(1000));
      expect(w % 16, 0);
    });

    test('caps each axis independently, preserving aspect', () {
      final (w, h) = SizeMath.targetDimensions(
        4000,
        2000,
        const VideoCompressConfig(
          maxWidth: 1920,
          alignment: DimensionAlignment.none,
        ),
      );
      expect(w, lessThanOrEqualTo(1920));
      // 2:1 source stays 2:1.
      expect((w / h - 2).abs(), lessThan(0.05));
    });

    test('only ever scales down — a cap above the source is a no-op', () {
      expect(
        SizeMath.targetDimensions(
          640,
          480,
          const VideoCompressConfig(maxWidth: 4000, maxHeight: 4000),
        ),
        (640, 480),
      );
    });

    test('alignment none still yields even dimensions', () {
      final (w, h) = SizeMath.targetDimensions(
        1001,
        667,
        const VideoCompressConfig(
          maxWidth: 1000,
          alignment: DimensionAlignment.none,
        ),
      );
      expect(w.isEven, isTrue);
      expect(h.isEven, isTrue);
    });

    test('degenerate source dimensions pass through instead of throwing', () {
      expect(SizeMath.targetDimensions(0, 0, const VideoCompressConfig()), (
        0,
        0,
      ));
      expect(
        SizeMath.targetDimensions(
          -1,
          -1,
          const VideoCompressConfig(maxWidth: 100),
        ),
        (-1, -1),
      );
    });

    test('never returns a zero side for a tiny cap', () {
      final (w, h) = SizeMath.targetDimensions(
        1920,
        1080,
        const VideoCompressConfig(maxWidth: 8),
      );
      expect(w, greaterThan(0));
      expect(h, greaterThan(0));
    });
  });
}
